src/sentry/sentry_metrics/multiprocess.py

```diff
@@ -0,0 +1,334 @@
+
+    @property
+    def messages(self):
+        return self.__messages
+
+    def append(self, message: Message[TPayload]) -> None:
+        self.__messages.append(message)
+
+    def ready(self) -> bool:
+        if len(self.messages) >= self.__max_batch_size:
+            return True
+        elif time.time() >= self.__deadline:
+            return True
+        else:
+            return False
+
+
+class BatchMessages(ProcessingStep):
+    def __init__(
+        self,
+        next_step: ProcessingStrategy[TProcessed],
+        max_batch_time: float,
+        max_batch_size: int,
+    ):
+        self.__max_batch_size = 2
+        self.__max_batch_time = max_batch_time
+
+        self.__next_step = next_step
+        self.__batch: Optional[MetricsBatchBuilder] = None
+        self.__closed = False
+
+    def poll(self) -> None:
+        assert not self.__closed
+
+        self.__next_step.poll()
+
+        if self.__batch and self.__batch.ready():
+            self.__flush()
+
+    def submit(self, message: Any) -> None:
+        if self.__batch is None:
+            self.__batch = MetricsBatchBuilder(self.__max_batch_size, self.__max_batch_time)
+
+        self.__batch.append(message)
+
+    def __flush(self) -> None:
+        if not self.__batch:
+            return
+        last = self.__batch.messages[-1]
+
+        new_message = Message(last.partition, last.offset, self.__batch.messages, last.timestamp)
+        self.__next_step.submit(new_message)
+        self.__batch = None
+
+    def terminate(self) -> None:
+        self.__closed = True
+
+    def close(self) -> None:
+        self.__closed = True
+        pass
+
+    def join(self, timeout: Optional[float] = None) -> None:
+        self.__next_step.join(timeout)
+
+
+class ProduceStep(ProcessingStep):
+    """
+    Step that produces to the snuba-metrics topic, collecting the futures returned by
+    the producer. Continously checks to see if futures are done, and once that's the case
+    can commit up to the last future that is done.
+    """
+
+    def __init__(self, commit_function: Callable[[Mapping[Partition, Position]], None]) -> None:
+        snuba_metrics = settings.KAFKA_TOPICS[settings.KAFKA_SNUBA_METRICS]
+        snuba_metrics_producer = KafkaProducer(
+            kafka_config.get_kafka_producer_cluster_options(snuba_metrics["cluster"]),
+        )
+        self.__producer = snuba_metrics_producer
+        self.__producer_topic = settings.KAFKA_TOPICS[settings.KAFKA_SNUBA_METRICS].get(
+            "topic", "snuba-metrics"
+        )
+        self.__commit_function = commit_function
+
+        self.__futures = []
+        self.__closed = False
+
+    def poll(self) -> None:
+        # TODO: continously check if futures are done i.e. future.done()
+        # and if so then we know we can commit
+        pass
+
+    def submit(self, outer_message: Message[TPayload]) -> None:
+        assert not self.__closed
+
+        for message in outer_message.payload:
```

Comment:
please keep a maximum size for the futures list. If the list is full raise `MessageRejected` this is interpreted as back pressure by the StreamProcessor, which at that point backs off (which means obnoxiously retries until it gets the message through)